test/backends.py

```diff
@@ -0,0 +1,217 @@
+__copyright__ = """
+Copyright (C) 2020 University of Illinois Board of Trustees
+Copyright (C) 2021 Esteban Cisneros
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import cantera as ct
+import numpy as np
+import importlib
+import pathlib
+import typing
+
+import pyrometheus
+
+from collections.abc import Iterable
+from abc import abstractmethod
+
+try:
+    import jax
+    import jax.numpy as jnp
+except ImportError:
+    jnp = None
+else:
+    jax.config.update("jax_enable_x64", True)
+
+
+class Backend:
+    def __init__(self, pyro):
+        self.pyro = pyro
+
+    @staticmethod
+    @abstractmethod
+    def get_name():
+        pass
+
+    @abstractmethod
+    def fix_arg(self, x):
+        return x
+
+    @abstractmethod
+    def fix_result(self, result):
+        return result
+
+    @staticmethod
+    @abstractmethod
+    def generate_code(sol: ct.Solution, name: str):
+        pass
+
+    @staticmethod
+    @abstractmethod
+    def extract_interface(module, mechname: str, user_np):
+        return module
+
+
+class PythonBackend(Backend):
```

Comment:
I like this solution, but I don't understand why it doesn't include the numpy-like module logic, deferred until line 139. Ideally, we'd include that logic here, especially as we generalize the numpy-like module to an array context (see [here](https://github.com/pyrometheus/pyrometheus/blob/arraycontext/pyrometheus/make_array.py)).

Reply:
I saw the other comment about this and I think I'm misunderstanding what you mean since the `JaxPyroWrapper` class is declared within a method of the `PythonBackend`. Perhaps you dislike the nesting? It could be placed outside with a little indirection

Reply:
Ah ok, my bad. I see that it is inside the class. I must've misinterpreted what I saw in my original review.